compiler/src/vm/heap/object.rs

```diff
@@ -73,38 +72,57 @@ impl Struct {
         }
         s
     }
-    fn insert(&mut self, heap: &Heap, key: Pointer, value: Pointer) {
-        let bucket = self.buckets.entry(key.hash(heap)).or_default();
-        for (existing_key, existing_value) in bucket.iter_mut() {
+    /// If the struct contains the key, returns the index of its field.
+    /// Otherwise, returns the index of where the key would be inserted.
+    fn index_of_key(&self, heap: &Heap, key: Pointer) -> Result<usize, usize> {
+        let hash = key.hash(heap);
+        let index_of_hash = self
+            .fields
+            .binary_search_by_key(&hash, |(hash, _, _)| *hash)?;
+        let going_right = self.fields[index_of_hash..]
+            .iter()
+            .enumerate()
+            .take_while(|(_, (existing_hash, _, _))| *existing_hash == hash)
+            .map(|(i, (_, key, _))| (index_of_hash + i, key));
+        let going_left = self.fields[0..index_of_hash]
+            .iter()
+            .rev()
+            .enumerate()
+            .take_while(|(_, (existing_hash, _, _))| *existing_hash == hash)
+            .map(|(i, (_, key, _))| (index_of_hash - i - 1, key));
```

Comment:
You can simplify this by searching for `(hash, Pointer::null)`: This always finds the left-most index for the given hash. This way, you only have to search in one direction from there.